### PR TITLE
Add project serialization

### PR DESCRIPTION
**Pull request by rschamp**
### Resolves
Requires https://github.com/LLK/scratch-storage/pull/11. Tests will fail until this is merged and published.

Resolves https://github.com/LLK/scratch-vm/issues/550 and https://github.com/LLK/scratch-gui/issues/329
Also is a start on https://github.com/LLK/scratch-vm/issues/194 (not final until we have a parser).

### Proposed Changes

This is a merge of the work done by @TheBrokenRail, @morantsur, and @thisandagain into the current VM.
- Add JSON serialization of project data, to serve as a first iteration of SB3 JSON
- Add sb3 json de/serializer
- Refactor sb2 loading to use a sb2 deserializer
- Stop storing costume and sound data in the VM state. 

### Reason for Changes

- Gives us a way to save/load sb3 files.
- Improve GUI performance by not sending costume and sound data in target updates. The GUI wants to compare the targets, and walking ArrayBuffers is not very fun

/cc @thisandagain @paulkaplan

### CODE REVIEW

File: src/virtual-machine.js [hunk 1]
```diff
@@ -180,11 +166,78 @@ class VirtualMachine extends EventEmitter {
...
+     * @returns {Promise} Promise that resolves after the project has loaded
+     */
+    fromJSON (json) {
+        // Clear the current runtime
+        this.clear();
+
+        // Validate & parse
+        if (typeof json !== 'string') return;
+        json = JSON.parse(json);
+        if (typeof json !== 'object') return;
```
Comment (cwillisf) [inline src/virtual-machine.js]: These early-outs should probably log an error or something (same in `addSprite2`)

File: src/virtual-machine.js [hunk 2]
```diff
@@ -180,11 +166,78 @@ class VirtualMachine extends EventEmitter {
...
+        return deserializer.deserialize(json, this.runtime).then(targets => {
+            this.clear();
+            for (let n = 0; n < targets.length; n++) {
+                if (targets[n] !== null) {
+                    this.runtime.targets.push(targets[n]);
+                    targets[n].updateAllDrawableProperties();
+                }
+            }
+            // Select the first target for editing, e.g., the first sprite.
+            this.editingTarget = this.runtime.targets[1];
```
Comment (cwillisf) [inline src/virtual-machine.js]: There's no requirement for a project to include a sprite. I'd recommend something equivalent to:
`this.editingTarget = this.runtime.targets[targets.length > 1 ? 1 : 0];`

That way the stage is selected if there are no sprites. If there isn't even a stage then the editing target will be `undefined`, which I guess is the best we can hope for in that situation...

**Review by cwillisf — CHANGES_REQUESTED** [review]

**Comment by cwillisf** [comment]
BTW: what happened with the history on this branch?

**Comment by rschamp** [comment]
That's the actual history :)  I will squash when I merge.

**Comment by rschamp** [comment]
@cwillisf Updated

**Review by cwillisf — APPROVED** [review]

**Pull request merged**